frb_codegen/src/generator/dart/ty_boxed.rs

```diff
@@ -17,18 +17,40 @@ impl TypeDartGeneratorTrait for TypeBoxedGenerator<'_> {
                 self.ir.inner.safe_ident(),
             )
         });
+
+        let contains_opacity = self
+            .ir
+            .inner
+            .distinct_types(self.context.ir_file)
+            .iter()
+            .any(IrType::is_opaque);
+
         Acc {
             io: Some(as_primitive.clone().unwrap_or_else(|| {
                 if self.ir.inner.is_array() {
                     format!("return api2wire_{}(raw);", self.ir.inner.safe_ident(),)
+                } else if contains_opacity {
+                    format!(
+                        "final ptr = inner.new_{ident}_{context}();
+                            try {{
+                                _api_fill_to_wire_{inner}(raw, ptr.ref);
+                            }} catch(e) {{
+                                inner.drop_{ident}_{context}(ptr);
+                                rethrow;
+                            }}
+                            return ptr;",
+                        ident = self.ir.safe_ident(),
+                        context = self.context.config.block_index,
+                        inner = self.ir.inner.safe_ident(),
+                    )
                 } else {
                     format!(
-                        "final ptr = inner.new_{}_{}();
-                        _api_fill_to_wire_{}(raw, ptr.ref);
-                        return ptr;",
-                        self.ir.safe_ident(),
-                        self.context.config.block_index,
-                        self.ir.inner.safe_ident(),
+                        "final ptr = inner.new_{ident}_{context}();
+                            _api_fill_to_wire_{inner}(raw, ptr.ref);
+                            return ptr;",
+                        ident = self.ir.safe_ident(),
+                        context = self.context.config.block_index,
+                        inner = self.ir.inner.safe_ident(),
```

Comment:
If I understand correctly, the logic is: "if contains_opacity, add extra try-catch to drop value".

Then, what about: Instead of contains_opacity, maybe something like children_is_throwable_when_filling